arch/x86/core/intel64/locore.S

```diff
@@ -0,0 +1,599 @@
+    IDT(252, INTR, 1); IDT(253, INTR, 1); IDT(254, INTR, 1); IDT(255, INTR, 1)
+
+idt48:
+	.word (idt48 - idt - 1)
+	.long idt
+
+#define EXCEPT_CODE(nr) vector_ ## nr: pushq $nr; jmp except
+#define EXCEPT(nr)	vector_ ## nr: pushq $0; pushq $nr; jmp except
+
+except: /*
+	 * finish struct NANO_ESF on stack. 'vector' .. 'ss' are
+	 * already there from hardware trap and EXCEPT_*() stub.
+	 */
+
+	pushq %r15
+	pushq %r14
+	pushq %r13
+	pushq %r12
+	pushq %r11
+	pushq %r10
+	pushq %r9
+	pushq %r8
+	pushq %rdi
+	pushq %rsi
+	pushq %rbp
+	pushq %rdx
+	pushq %rcx
+	pushq %rbx
+	pushq %rax
+
+	movl $_NANO_ERR_CPU_EXCEPTION, %edi
+	movq %rsp, %rsi
+	call z_NanoFatalErrorHandler
+
+	hlt /* should not return, but just in case .. */
+
+EXCEPT      ( 0); EXCEPT      ( 1); EXCEPT      ( 2); EXCEPT      ( 3)
+EXCEPT      ( 4); EXCEPT      ( 5); EXCEPT      ( 6); EXCEPT      ( 7)
+EXCEPT_CODE ( 8); EXCEPT      ( 9); EXCEPT_CODE (10); EXCEPT_CODE (11)
+EXCEPT_CODE (12); EXCEPT_CODE (13); EXCEPT_CODE (14); EXCEPT      (15)
+EXCEPT      (16); EXCEPT_CODE (17); EXCEPT      (18); EXCEPT      (19)
+EXCEPT      (20); EXCEPT      (21); EXCEPT      (22); EXCEPT      (23)
+EXCEPT      (24); EXCEPT      (25); EXCEPT      (26); EXCEPT      (27)
+EXCEPT      (28); EXCEPT      (29); EXCEPT      (30); EXCEPT      (31)
+
+/*
+ * When we arrive at 'irq' from one of the IRQ(X) stubs,
+ * we're on the "freshest" IRQ stack and it contains:
+ *
+ *   SS
+ *   RSP
+ *   RFLAGS
+ *   CS
+ *   RIP
+ *   (vector number - IV_IRQS) <-- RSP points here
+ *   RSI <-- we push this on entry
+ */
+
+.globl x86_irq_funcs	/* see irq_manage.c .. */
+.globl x86_irq_args	/* .. for these definitions */
+
+irq:
+	pushq %rsi
+	movq $_kernel, %rsi
+
+	/*
+	 * Bump the IRQ nesting count and move to the next IRQ stack.
+	 * That's sufficient to safely re-enable interrupts, so if we
+	 * haven't reached the maximum nesting depth yet, do it.
+	 */
+
+	incl _kernel_offset_to_nested(%rsi)
+	addq $CONFIG_ISR_STACK_SIZE, ist1
+	cmpl $CONFIG_ISR_DEPTH, _kernel_offset_to_nested(%rsi)
+	jz 1f
+	sti
+1:	cmpl $1, _kernel_offset_to_nested(%rsi)
+	je irq_enter_unnested
+
+	/*
+	 * if we're a nested interrupt, we have to dump the state to the
+	 * stack. we play some games here to re-arrange the stack thusly:
+	 *
+	 * SS RSP RFLAGS CS RIP RAX RSI RCX RDX RBX
+	 * RDI RBP R8 R9 R10 R11 R12 R13 R14 R15
+	 * X86_FXSAVE_SIZE bytes of SSE data <-- RSP points here
+	 *
+	 * note that the final value of RSP must be 16-byte aligned here,
+	 * both to satisfy FXSAVE/FXRSTOR but also to honor the C ABI.
+	 */
+
+irq_enter_nested: /* Nested IRQ: dump register state to stack. */
+	pushq %rcx
+	movq 16(%rsp), %rcx /* RCX = vector */
+	movq %rax, 16(%rsp) /* looks like we pushed RAX, not the vector */
+	pushq %rdx
+	pushq %rbx
+	pushq %rdi
+	pushq %rbp
+	pushq %r8
+	pushq %r9
+	pushq %r10
+	pushq %r11
+	pushq %r12
+	pushq %r13
+	pushq %r14
+	pushq %r15
```

Comment:
This duplicates the exception save code above in except(), more or less needlessly.   The original implementation had a single handler (at the cost of an instruction or three to handle the error code on some exceptions).

Reply:
The exception path may look quite similar, but it's deliberately separated from the IRQ path, because they're different things.  At present, exceptions are all treated as fatal, and all the pushing there is really just to create an exception frame for the handler to dump.  So they're superficially nearly identical, but they're conceptually different, and will diverge over time, so they're uncoupled.